src/Uno.UI/Helpers/WeakEvents/WeakEventManager.cs

```diff
@@ -3,133 +3,131 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Reflection;
 using System.Runtime.CompilerServices;
 
-// Mostly from https://github.com/dotnet/maui/blob/c92d44c68fe81c57ef8caec2506e6788309b8ff4/src/Core/src/WeakEventManager.cs
-// But adjusted a little bit
+namespace Uno.UI.Helpers;
 
-namespace Uno.UI.Helpers
+internal sealed class WeakEventManager
 {
-	internal sealed class WeakEventManager
-	{
-		private readonly Dictionary<string, List<Subscription>> _eventHandlers = new(StringComparer.Ordinal);
+	private readonly ConditionalWeakTable<object, Dictionary<string, List<Action>>> _handlersPerTarget = new();
+	private Dictionary<string, List<Action>>? _staticHandlers;
+	private string? _enumeratingHandlersOfEvent;
+	private Stack<string>? _enumeratingHandlersOfEventStack;
 
-		public void AddEventHandler(Action? handler, [CallerMemberName] string eventName = "")
+	public void AddEventHandler(Action? handler, [CallerMemberName] string eventName = "")
+	{
+		if (string.IsNullOrEmpty(eventName))
 		{
-			if (string.IsNullOrEmpty(eventName))
-				throw new ArgumentNullException(nameof(eventName));
-
-			if (handler == null)
-				throw new ArgumentNullException(nameof(handler));
-
-			AddEventHandler(eventName, handler.Target, handler.GetMethodInfo());
+			throw new ArgumentNullException(nameof(eventName));
 		}
 
-		public void HandleEvent(string eventName)
+		if (handler is null)
 		{
-			if (_eventHandlers.TryGetValue(eventName, out List<Subscription>? target))
-			{
-				for (int i = 0; i < target.Count; i++)
-				{
-					Subscription subscription = target[i];
-					bool isStatic = subscription.Subscriber == null;
-					if (isStatic)
-					{
-						// For a static method, we'll just pass null as the first parameter of MethodInfo.Invoke
-						subscription.Handler.Invoke(null, null);
-						continue;
-					}
-
-					object? subscriber = subscription.Subscriber?.Target;
-
-					if (subscriber == null)
-					{
-						// The subscriber was collected, so there's no need to keep this subscription around
-						target.Remove(subscription);
-					}
-					else
-					{
-						subscription.Handler.Invoke(subscriber, null);
-					}
-				}
-			}
+			// Allow handler to be nullable to not having to deal with nullability checks in callsites, but this indicates a bug.
+			Debug.Fail($"Got a null event handler for event {eventName}.");
```

Comment:
Debug.Assert() ?

Reply:
The assert is the condition above, if we reach this point we already failed!